js/src/index.js

```diff
@@ -553,6 +595,138 @@ class JupyterScatterView {
+
+      this[`tooltipContent${capitalContent}Channel`].appendChild(
+        this[`tooltipContent${capitalContent}ChannelName`]
+      );
+
+      this[`tooltipContent${capitalContent}ChannelBadge`] = createElementWithClass(
+        'div',
+        ['channel-badge', `${content}-value-badge`]
+      );
+      this[`tooltipContent${capitalContent}ChannelBadgeMark`] = createElementWithClass(
+        'div',
+        ['channel-badge-mark', `${content}-channel-badge-mark`]
+      );
+      this[`tooltipContent${capitalContent}ChannelBadgeBg`] = createElementWithClass(
+        'div',
+        ['channel-badge-bg', `${content}-channel-badge-bg`]
+      );
+      this[`tooltipContent${capitalContent}ChannelBadge`].appendChild(
+        this[`tooltipContent${capitalContent}ChannelBadgeMark`]
+      );
+      this[`tooltipContent${capitalContent}ChannelBadge`].appendChild(
+        this[`tooltipContent${capitalContent}ChannelBadgeBg`]
+      );
+      this[`tooltipContent${capitalContent}Channel`].appendChild(
+        this[`tooltipContent${capitalContent}ChannelBadge`]
+      );
+    }
+
+    this[`tooltipContent${capitalContent}Value`].appendChild(
+      this[`tooltipContent${capitalContent}ValueText`]
+    );
+
+    this[`tooltipContent${capitalContent}ValueHistogram`] = createHistogram(
+      TOOLTIP_HISTOGRAM_WIDTH,
+      TOOLTIP_HISTOGRAM_HEIGHT
+    );
+    this[`tooltipContent${capitalContent}Value`].appendChild(
+      this[`tooltipContent${capitalContent}ValueHistogram`].element
+    );
+
+    const histogram = this.model.get(`${content}_histogram`) || this.model.get('tooltip_contents_non_visual_info')[content]?.histogram;
+    const scale = this.model.get(`${content}_scale`) || this.model.get('tooltip_contents_non_visual_info')[content]?.scale
+
+    this[`tooltipContent${capitalContent}ValueHistogram`].init(
+      histogram,
+      scale === 'categorical',
+    );
+
+    this[`tooltipContent${capitalContent}Title`].textContent = toTitleCase(
+      this.model.get(`${content}_title`) || content || ''
+    );
+
+    this.tooltipContent.appendChild(this[`tooltipContent${capitalContent}Channel`]);
+    this.tooltipContent.appendChild(this[`tooltipContent${capitalContent}Title`]);
+    this.tooltipContent.appendChild(this[`tooltipContent${capitalContent}Value`]);
+  }
+
+  createTooltipContentsDomElements() {
+    this.tooltipContent.replaceChildren();
+
+    const contents = new Set(this.tooltipContentsAll);
+
+    // To ensure that visual contents are listed before non-visual contents,
+    // we're first going to iterate over the visual contents.
+    for (const content of TOOLTIP_ALL_VISUAL_CONTENTS) {
+      if (!contents.has(content)) continue;
+      contents.delete(content);
+      this.createTooltipContentDomElements(content);
+    }
+
+    // Next, we're going to iterate over the remaining non-visual contents
+    for (const content of contents) {
+      this.createTooltipContentDomElements(content);
+    }
```

Comment:
Yeah you're right. There are two bugs here and I believe they both cancel each other out. haha
1. At some point I was using Sets for `TOOLTIP_ALL_VISUAL_CONTENTS` but then decided to switch back to Maps
2. Initially I was using Sets for `tooltip_contents` and sets don't preserve the order between the Python and JS kernel. Since I didn't like that, I switched to lists.

Now given that I want the order of `tooltip_contents` to be preserved, we can actually remove lines 688-694. I don't think here's any reason to enforce visual properties to appear before non-visual properties. I really only did that because previously the properties would appear in random order all the time as the set order wasn't preserved (and appeared to be random on every widget render call)

What do you think?

Reply:
Sounds good! let's ditch these lines 😎